File: internal/store/review.go
package store

import (
	"database/sql"

	"github.com/xebia/ai-assisted-se-claude-exercise/internal/model"
)

type ReviewStore struct {
	db *sql.DB
}

func NewReviewStore(db *sql.DB) *ReviewStore {
	return &ReviewStore{db: db}
}

func (s *ReviewStore) ListByBook(bookID int) ([]model.Review, error) {
	rows, err := s.db.Query(
		`SELECT id, book_id, rating, review_text, created_at FROM reviews WHERE book_id = ?`, bookID,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var reviews []model.Review
	for rows.Next() {
		var r model.Review
		if err := rows.Scan(&r.ID, &r.BookID, &r.Rating, &r.ReviewText, &r.CreatedAt); err != nil {
			return nil, err
		}
		reviews = append(reviews, r)
	}
	return reviews, rows.Err()
}

func (s *ReviewStore) Create(bookID, rating int, reviewText string) (*model.Review, error) {
	res, err := s.db.Exec(
		`INSERT INTO reviews (book_id, rating, review_text) VALUES (?, ?, ?)`,
		bookID, rating, reviewText,
	)
	if err != nil {
		return nil, err
	}
	id, _ := res.LastInsertId()
	var r model.Review
	err = s.db.QueryRow(
		`SELECT id, book_id, rating, review_text, created_at FROM reviews WHERE id = ?`, id,
	).Scan(&r.ID, &r.BookID, &r.Rating, &r.ReviewText, &r.CreatedAt)
	return &r, err
}

File: bookstore/store/review.py
import sqlite3
from typing import Optional

from bookstore.model.review import Review


class ReviewStore:
    def __init__(self, db: sqlite3.Connection):
        self.db = db

    def list_by_book(self, book_id: int) -> list[Review]:
        rows = self.db.execute(
            "SELECT id, book_id, rating, review_text, created_at FROM reviews WHERE book_id = ?",
            (book_id,),
        ).fetchall()
        return [Review(**dict(r)) for r in rows]

    def create(self, book_id: int, rating: int, review_text: str) -> Review:
        cur = self.db.execute(
            "INSERT INTO reviews (book_id, rating, review_text) VALUES (?, ?, ?)",
            (book_id, rating, review_text),
        )
        self.db.commit()
        row = self.db.execute(
            "SELECT id, book_id, rating, review_text, created_at FROM reviews WHERE id = ?",
            (cur.lastrowid,),
        ).fetchone()
        return Review(**dict(row))
